### PR TITLE
[WIP] Logging: language dependent logging

### PR DESCRIPTION
**Pull request by raduml97**
Allows for different target languages to log different types of content.

### CODE REVIEW

File: Gillian-C/lib/External.ml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 open Gillian.Gil_syntax
 open Gillian.General
+module L = Gillian.Logging.Default
```
Comment (giltho) [inline Gillian-C/lib/External.ml:3]: Not liking this, if you change the logger you change everywhere
Maybe create a Logging module in `Gillian-C` that just contains `include Gillian.Logging.Default`

Comment (giltho) [reply Gillian-C/lib/External.ml:3]: Then you can also just call `Logging` everywhere, which is nicer too

File: GillianCore/logging/logging.ml [hunk 2]
```diff
@@ -1,129 +1,250 @@
...
-let log lvl msgf =
-  if Mode.should_log lvl then
-    let report = ReportBuilder.info "" (Debug msgf) () in
-    Reporter.log report
+  let reporters =
+    [ new Reporter.file_reporter (); new Reporter.database_reporter () ]
+end
 
-let normal msgf = log Normal msgf
+module type S = sig
```
Comment (giltho) [inline GillianCore/logging/logging.ml:11]: One solution you could adopt to reduce duplication is creating a `reporter_S.mli` which just contains the signature.
In the dune file you will need to add `(modules_without_implementation reporter_S)`
Then you should be able to say things like
```ocaml
module Make(P: TargetLang.S) : Reporter_S
module Default : Reporter_S
```

I find `Reporter_S` a little less elegant than `Reporter.S`, but code duplication is difficult to maintain

**Comment by giltho** [comment]
Ok, I think there's a tiny thing wrong with the design we've chosen
If I want to report something during the execution of an action, I need my logger instantiated. But to instantiate the logger I need the types of my memory model to be defined and I need to have instantiated my memory model

One solution is probably passing the reporters to the actions.

My second issue is, you're importing `Reporter.Default` everywhere in the code.
If I change my reporter module, do I need to make every module in Gillian a functor parametrized by the Reporter module ? That can't work. There has to be a better solution. I need to think about it.

**Comment by raduml97** [comment]
Will reopen a new PR for this.

**Pull request closed without merging**